### DISCUSSION TITLE
UE5 HELP REQUESTED: Send ECEFAngularVel properly to JSBSim plugin so the plane rotates at the same angular velocity as the ground.

### DISCUSSION THREAD

**Discussion by miju70** [JSBSim-Team/jsbsim#1158] [Q&A]
I'm trying to takeoff and land planes over ships that are moving in an Unreal Engine 5 level, in order to use JSBSim flight dynamics I have found there is a function inside UEGroundCallback which gets the angular velocity on ECEF coordinates and send it to accelerations.cpp to zero the relative angular velocity between the plane and the terrain:

```c++
double UEGroundCallback::GetAGLevel(const FGLocation& location, FGLocation& contact, FGColumnVector3& normal, FGColumnVector3& vel, FGColumnVector3& angularVel) const
{
	vel.InitMatrix();
	angularVel.InitMatrix();

	FVector ECEFLocation = FVector(location(1), location(2), location(3)) * FEET_TO_METER;
	FVector ECEFContactPoint;
	FVector Normal;
	FVector ECEFVel;
	FVector ECEFAngularVel;
	double AGDistance = MovementComponent->GetAGLevel(ECEFLocation, ECEFContactPoint, Normal, ECEFVel, ECEFAngularVel) * METER_TO_FEET;

	contact.SetEllipse(20925646.32546, 20855486.5951); // Important to set these values, otherwise the contact location is not valid, and the FGLocation::FGLocation(FGColumnVector3) don't do it!
	contact = FGColumnVector3(ECEFContactPoint.X, ECEFContactPoint.Y, ECEFContactPoint.Z)*METER_TO_FEET; // The = operators sets ECEF Coordinates

	normal = FGColumnVector3(Normal.X, Normal.Y, Normal.Z);
	
	vel = FGColumnVector3(ECEFVel.X, ECEFVel.Y, ECEFVel.Z) * METER_TO_FEET;
	angularVel = FGColumnVector3(ECEFAngularVel.X, ECEFAngularVel.Y, ECEFAngularVel.Z);

	return AGDistance;
 }
```

Well in order to get the ship angular velocity from the plane, I have created a component called MovingGroundComponent and have attached it to the ship, this component provides the angular velocity of its owner the next way:

```c++
    FQuat CurrentRotationQuat = GetOwner()->GetActorQuat();
    FQuat DeltaQuat = CurrentRotationQuat * PreviousRotationQuat.Inverse();

    FVector RotationAxis;
    float RotationAngle;
    DeltaQuat.ToAxisAndAngle(RotationAxis, RotationAngle);

    float RotationAngleDegrees = FMath::RadiansToDegrees(RotationAngle);

    FVector AngularVelocityDegrees = (RotationAxis * RotationAngleDegrees) / DeltaTime;

    FVector AngularVelocityRadians;
    AngularVelocityRadians.X = AngularVelocityDegrees.X * (PI / 180);
    AngularVelocityRadians.Y = AngularVelocityDegrees.Y * (PI / 180);
    AngularVelocityRadians.Z = AngularVelocityDegrees.Z * (PI / 180);

    AngularVelocityRadians.X = -AngularVelocityRadians.X;
    AngularVelocityRadians.Y = -AngularVelocityRadians.Y;

    const FVector CurrentLocationUE = GetOwner()->GetActorLocation();
    FGeographicCoordinates CurrentLocationGEO;
    GeoReferencingSystem->EngineToGeographic(CurrentLocationUE, CurrentLocationGEO);

    float Longitude = CurrentLocationGEO.Longitude;
    float Latitude = CurrentLocationGEO.Latitude;

    FMatrix ENUtoECEF;
    ENUtoECEF.SetIdentity();

    ENUtoECEF.M[0][0] = -FMath::Sin(Longitude);
    ENUtoECEF.M[0][1] = FMath::Cos(Longitude);
    ENUtoECEF.M[0][2] = 0;

    ENUtoECEF.M[1][0] = -FMath::Cos(Longitude) * FMath::Sin(Latitude);
    ENUtoECEF.M[1][1] = -FMath::Sin(Longitude) * FMath::Sin(Latitude);
    ENUtoECEF.M[1][2] = FMath::Cos(Latitude);

    ENUtoECEF.M[2][0] = -FMath::Cos(Longitude) * FMath::Cos(Latitude);
    ENUtoECEF.M[2][1] = -FMath::Sin(Longitude) * FMath::Cos(Latitude);
    ENUtoECEF.M[2][2] = -FMath::Sin(Latitude);

    FVector AngularVelocityECEF = ENUtoECEF.TransformVector(AngularVelocityRadians);
    AngularVelocity = AngularVelocityECEF;

    PreviousRotationQuat = CurrentRotationQuat;
```

This last AngularVelocity variable is sent to JSBSim as ECEFAngularVel.

However, the relative rotation between plane and ship still not 0.

Has anyone tried to send angular velocity of the terrain from UE5 to JSBSim before?

**Comment by seanmcleod** [comment]
@miju70 I'm a bit confused, the code you show for `UEGroundCallback::GetAGLevel` above doesn't match the code in the repo, see below. In particular in the repo `angularVel` isn't calculated or updated, the only reference to it is `angularVel.InitMatrix();`. 

It looks like your code snippet's call to `MovementComponent->GetAGLevel()` is also different to what is in the repo.

https://github.com/JSBSim-Team/jsbsim/blob/8e4803b2cbae945fd62e84b1aca121c35edf5174/UnrealEngine/Plugins/JSBSimFlightDynamicsModel/Source/JSBSimFlightDynamicsModel/Private/UEGroundCallback.cpp#L16-L36

Do you have some updated/customised version?

**Reply by miju70** [reply]
Hey @seanmcleod!

Indeed, I have customize the code in order to send the angular velocity to JSBSim, on the original repo this velocity is completely ignored.

What I'm trying to figure out if is someone has achieved to comunicate the translation and rotation of the ground below the plane to JSBSim so the plane does not slide over it.

**Comment by seanmcleod** [comment]
> I have found there is a function inside UEGroundCallback which gets the angular velocity on ECEF coordinates and send it to accelerations.cpp

Okay, your original comment made it sound like you had found some existing code, as opposed to this being new customised code you've developed.

> on the original repo this velocity is completely ignored.

Correct, that's where I first looked, in particular in the landing gear code, where the angular velocity is ignored - `dummy`.

https://github.com/JSBSim-Team/jsbsim/blob/8e4803b2cbae945fd62e84b1aca121c35edf5174/src/models/FGLGear.cpp#L287-L298

The `terrainVel` though is used, see line 355:

https://github.com/JSBSim-Team/jsbsim/blob/8e4803b2cbae945fd62e84b1aca121c35edf5174/src/models/FGLGear.cpp#L351-L356

**Comment by seanmcleod** [comment]
@miju70 FlightGear supports take-offs and landings on aircraft carriers. I'm not aware of them making any specific changes to JSBSim in order to support this, in which case making use of the `terrainVel` only without the angular velocity may well be enough. You could check the FlightGear source code to double-check.

**Comment by seanmcleod** [comment]
Hmm, taking a quick look at the history for `FGLGear.cpp` I see @bcoconni pushed a commit back in Oct 2010 - "Patch from Bertrand to fix ground reactions on a turning carrier" - https://github.com/JSBSim-Team/jsbsim/commit/346235fd3f1eaed954b8947a70be08a0c2d5b7a3 

Which added support for `angularVel`. Will take a closer look.

**Comment by seanmcleod** [comment]
> on the original repo this velocity is completely ignored.

When I first looked into this, see my comment - https://github.com/JSBSim-Team/jsbsim/discussions/1158#discussioncomment-10635938 I looked at the gear contact code and sure enough in that code the angular velocity is ignored, only the translational velocity of the ground is used.

However, on further inspection after finding @bcoconni's commit I mentioned above I see that JSBSim does make use of the ground's angular velocity.

First off, it's queried in `FGPropagate.cpp` during each timestep for the aircraft's cg.

https://github.com/JSBSim-Team/jsbsim/blob/8e4803b2cbae945fd62e84b1aca121c35edf5174/src/models/FGPropagate.cpp#L592-L598

Then it's sent as an input to the accelerations code.

https://github.com/JSBSim-Team/jsbsim/blob/8e4803b2cbae945fd62e84b1aca121c35edf5174/src/FGFDMExec.cpp#L614

Where, if at least one gear is in contact with the ground.

https://github.com/JSBSim-Team/jsbsim/blob/8e4803b2cbae945fd62e84b1aca121c35edf5174/src/models/FGAccelerations.cpp#L241-L242

Then the ground's angular velocity is finally used.

https://github.com/JSBSim-Team/jsbsim/blob/8e4803b2cbae945fd62e84b1aca121c35edf5174/src/models/FGAccelerations.cpp#L273-L274

**Comment by seanmcleod** [comment]
@miju70 any feedback?

**Reply by jijnt12345** [reply]
How to modify "LocalTerrainVelocity"

**Comment by seanmcleod** [comment]
You implement the ground callback, and if for example you determine that callback is for a point above an object like say an aircraft carrier, then you return the aircraft carrier's velocity as the `LocalTerrainVelocity`.

**Reply by jijnt12345** [reply]
Thank you for your answer，Do I only need to assign the carrier's linear velocity to LocalTerrainVelocity in FGPropagate.h every frame？

But LocalTerrainVelocity in FGPropagate.h is private ,Do I need to modify the JSBSim source code and recompile the plugin for UE5?

**Comment by seanmcleod** [comment]
No, take a look at comment - https://github.com/JSBSim-Team/jsbsim/discussions/1158#discussioncomment-10635753 and in particular the `UEGroundCallback::GetAGLevel` implementation that I showed. Note that the 4th argument `FGColumnVector3& vel` is passed by reference by JSBSim.

So all you need to do is set it. JSBSim will then make use of it, no need to modify any JSBSim source code.

Take note of the angular terrain velocity, the 5th argument. The same applies to this.

**Reply by jijnt12345** [reply]
Thanks,But I still have some issues. In the tick function, I get the carrier's position each frame and subtract the previous frame's position from the current frame's position to assign to vel. Regardless of whether I use UE5 units or ECEF, the aircraft does not move together with the carrier.

```c++
void UJSBSimMovementComponent::CalculateCarrierVelocity()
{
	TerrianVelocityFromCarrier = CarrierCurrentPosition-CarrierPastPosition;
	CarrierPastPosition = CarrierCurrentPosition;
}

double UEGroundCallback::GetAGLevel(const FGLocation& location, FGLocation& contact, FGColumnVector3& normal, FGColumnVector3& vel, FGColumnVector3& angularVel) const
{
	//vel.InitMatrix();
	vel = FGColumnVector3(MovementComponent->TerrianVelocityFromCarrier.X,MovementComponent->TerrianVelocityFromCarrier.Y,MovementComponent->TerrianVelocityFromCarrier.Z);
	angularVel.InitMatrix();

	FVector ECEFLocation = FVector(location(1), location(2), location(3)) * FEET_TO_METER;
	FVector ECEFContactPoint;
	FVector Normal;
	double AGDistance = MovementComponent->GetAGLevel(ECEFLocation, ECEFContactPoint, Normal) * METER_TO_FEET;

	contact.SetEllipse(20925646.32546, 20855486.5951); // Important to set these values, otherwise the contact location is not valid, and the FGLocation::FGLocation(FGColumnVector3) don't do it!
	contact = FGColumnVector3(ECEFContactPoint.X, ECEFContactPoint.Y, ECEFContactPoint.Z)*METER_TO_FEET; // The = operators sets ECEF Coordinates
	normal = FGColumnVector3(Normal.X, Normal.Y, Normal.Z);
	return AGDistance;
}
```

**Reply by jijnt12345** [reply]
Additionally, the aircraft has a change in position, but its direction of motion and speed magnitude are not consistent with those of the carrier

**Comment by seanmcleod** [comment]
Yep, but note that I also mentioned the terrain's angular velocity.

Also look at my other previous comment - https://github.com/JSBSim-Team/jsbsim/discussions/1158#discussioncomment-10648677 where I mention how JSBSim makes use of the terrain's angular velocity. Plus the explicit comment in the source code that I listed:

`// Zeroes out the relative movement between the aircraft and the ground `.

**Reply by jijnt12345** [reply]
I found that the terrain angular velocity is being calculated on its own and is quite accurate, so no additional modifications are needed.

**Comment by seanmcleod** [comment]
So are you saying that everything is working now?

**Reply by jijnt12345** [reply]
The angular velocity matches well, but there is always a significant difference in linear velocity

**Reply by jijnt12345** [reply]
Uploading JF17.mp4…

**Comment by seanmcleod** [comment]
From the JSBSim source code snippets I listed in comments above you'll note `in.Tec2b` is used for both the linear velocity and angular velocity that you return from your ground callback.

```c++
vBodyWhlVel += in.UVW - in.Tec2b * terrainVel; 
wdot += (in.vPQR - in.Tec2b * in.TerrainAngularVel) / dt;
```

Where:

```c++
  /** Retrieves the ECEF-to-body transformation matrix.
      @return a reference to the ECEF-to-body transformation matrix.  */
  const FGMatrix33& GetTec2b(void) const { return Tec2b; }
```

**Reply by jijnt12345** [reply]
Is the TerrainVelocity unit in ECEF or Unreal Engine 5 units? I suspect there might be a unit mismatch

**Comment by seanmcleod** [comment]
That was my point, JSBSim is expecting ECEF based velocities. And in terms of linear velocity JSBSim is expecting feet per second, whereas it looks like UE uses meters etc. 

Your JF-17 MP4 file, or link to it didn't work in your comment above.

**Reply by jijnt12345** [reply]
It was indeed a unit issue. After adjusting the units, everything is now working fine.

**Comment by seanmcleod** [comment]
Are @miju70 and you working together on the same project?

To calculate the terrain's (ship's) angular velocity are you using the code that @miju70 showed in the original post above?

**Reply by jijnt12345** [reply]
No, it's not the same project. I haven't made any additions or modifications to the angular velocity logic; I'm using the original plugin logic.

**Comment by seanmcleod** [comment]
But as I pointed out to @miju70 the original plugin logic doesn't compute the angular velocity at all.

https://github.com/JSBSim-Team/jsbsim/blob/8e4803b2cbae945fd62e84b1aca121c35edf5174/UnrealEngine/Plugins/JSBSimFlightDynamicsModel/Source/JSBSimFlightDynamicsModel/Private/UEGroundCallback.cpp#L16-L36

**Comment by seanmcleod** [comment]
@jijnt12345 the reason I'm asking about your implementation of the carrier rotation is so that if someone else has this same issue months later and does a search and comes across this discussion then they can see an example of some working code that they could use.

**Reply by jijnt12345** [reply]
I didn't perform any calculations for rotation because I found that the JSBSim native plugin can directly implement the aircraft's follow the carrier's tilt and rotation.

**Comment by seanmcleod** [comment]
Okay, so the only change you made to `UEGroundCallback::GetAGLevel` was to return the linear velocity of your aircraft carrier?

Just a bit surprised, given @miju70's original issue, plus @bcoconni's comment at the time he added support for the rotational velocity for the terrain to JSBSim:

> Patch from Bertrand to fix ground reactions on a turning carrier
> // Zeroes out the relative movement between the aircraft and the ground 

Maybe your aircraft carrier isn't turning, or maybe you immediately start a take-off getting off the carrier before it becomes an issue?

**Reply by jijnt12345** [reply]
I just tested it, and there was still some desynchronization between the aircraft wheels and the carrier during rough seas. I plan to add angular velocity calculations. What is the unit for angular velocity in JSBSim?

**Comment by seanmcleod** [comment]
Rad/s.

Are you definitely also updating the aircraft carrier's pose for every JSBSim time step and not at a slower rate like the configured UE graphics update rate?

**Reply by jijnt12345** [reply]
Currently, I am still updating it in the Tick of UE5.
What is the order of pitch, yaw, and roll in FGColumnVector3()?

**Comment by miju70** [comment]
@jijnt12345 The way I have managed to send properly the linear velocity of the carrier to JSBSim is by creating a component of the carrier that does the velocity calculation and then JSBSim asks the ship for this velocity and send to the accelerations via `UJSBSimMovementComponent::GetAGLevel` inside JSBSimMovementComponent.cpp:

```c++
double UJSBSimMovementComponent::GetAGLevel(
	const FVector& StartECEFLocation,
	FVector& ECEFContactPoint,
	FVector& ECEFNormal,
	FVector& ECEFVel,
	FVector& ECEFAngularVel)
{
if (GetWorld()->LineTraceSingleByObjectType(HitResult, LineCheckStart, LineCheckEnd, ObjectParams, CollisionParams))
	{
			LineTraceImpactPoint = HitResult.ImpactPoint;
			LineTraceImpactNormal = HitResult.ImpactNormal;
			
			if (AActor* HitActor = HitResult.GetActor())
			{
				if (UJSBSimMovingGroundComponent* MovingGroundComponent =
					HitActor->FindComponentByClass<UJSBSimMovingGroundComponent>())
				{
					ECEFVel = MovingGroundComponent->GetVelocity();

					ECEFAngularVel = MovingGroundComponent->GetAngularVelocity();
				}
			}
	}
}
```

And this `GetVelocity` function does the calculation:

```c++
const FVector CurrentLocationUE = GetOwner()->GetActorLocation();

	FVector CurrentLocationECEF;
	
	GeoReferencingSystem->EngineToECEF(CurrentLocationUE, CurrentLocationECEF);
	
	Velocity = (CurrentLocationECEF - LastTickLocation) / DeltaTime;

	LastTickLocation = CurrentLocationECEF;
```

**Comment by miju70** [comment]
However, I'm still working on a solution for the angular velocity, I have an aproximation that works if the plane pivot is aligned with the carrier pivot on Z axis:

```c++
FQuat CurrentRotationQuat = GetOwner()->GetActorQuat();

    FQuat DeltaQuat = CurrentRotationQuat * PreviousRotationQuat.Inverse();
	
    FVector RotationAxis;
    float RotationAngle;
    DeltaQuat.ToAxisAndAngle(RotationAxis, RotationAngle);

    float RotationAngleDegrees = FMath::RadiansToDegrees(RotationAngle);

    FVector AngularVelocityDegrees = (RotationAxis * RotationAngleDegrees) / DeltaTime;
	
    AngularVelocityRadians.X = AngularVelocityDegrees.X * (PI / 180);
    AngularVelocityRadians.Y = AngularVelocityDegrees.Y * (PI / 180);
    AngularVelocityRadians.Z = AngularVelocityDegrees.Z * (PI / 180);

    AngularVelocityRadians.X = -AngularVelocityRadians.X;
    AngularVelocityRadians.Y = -AngularVelocityRadians.Y;

   AngularVelocity = - 2.25 * AngularVelocityRadians;

    PreviousRotationQuat = CurrentRotationQuat;
```

But I can't achieve to make it works if the plane is on the carrier corner, as I need to add an aditional lateral velocity that matches the pivots diference.

**Comment by seanmcleod** [comment]
@miju70 @jijnt12345 when you post C++ code snippets make use of Github's C++ formatting, you'll see I've been editing your posts to enable it so that it's easier to read for everyone.

Basically you use 3 back ticks followed by the coding language, c++ in this case, then paste your code followed by 3 back ticks again.

```sh
```c++
some c++ code...
some c++ code...
```  - finish with 3 back ticks
```

**Comment by seanmcleod** [comment]
> But I can't achieve to make it works if the plane is on the carrier corner

@miju70 you could take a look at how FlightGear calculates the angular velocity for an aircraft carrier.

**Comment by gallonmate** [comment]
Just a small note to be aware that Flightgear is GPL licensed code. If planning to release a UE5 product to the public, then I would avoid copying any Flightgear code. If not planning to release UE5 product to public, then it's okay to copy.
